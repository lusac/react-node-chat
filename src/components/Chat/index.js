import React from 'react'
import PropTypes from 'prop-types'

export default class Chat extends React.Component {
  constructor(props) {
    super()
    this.state = {
      msg: '',
      room: props.room || {},
      msgs: props.room.msgs || []
    }
  }

  componentDidUpdate(prevProps) {
    if (this.props.room.id !== prevProps.room.id) {
      this.updateRoomConnection()
    }
  }

  updateRoomConnection() {
    this.setState({
      msgs: this.props.room.msgs
    }, () => {
      this.props.socket.off('message').on('message', (msg) => {
        this.setState({
          msgs: [...this.state.msgs, msg]
        })
      })
    })
  }

  onMsgChange(e) {
    this.setState({
      msg: e.target.value
    })
  }

  onMsgSubmit(e) {
    e.preventDefault()
    this.props.socket.emit('message', {
      roomID: this.props.room.id,
      msg: this.state.msg
    })
    this.setState({
      msg: ''
    })
  }

  render() {
    return (
      <div className="chat-room">
        <div className="chat-room__messages">
          {!this.props.room.id && <span>Nenhum chat selecionado</span>}
          {this.props.room.id && this.state.msgs.map((msg, i) => <div key={i}>{msg}</div> )}
        </div>

        {this.props.room.id &&
          <div className="chat-room__input__wrapper">
            <form onSubmit={this.onMsgSubmit.bind(this)}>
              <input
                placeholder="escreva uma mensagem"
                className="chat-room__input"
                value={this.state.msg}
                onChange={this.onMsgChange.bind(this)} />
              {/* <input type="submit" value="Submit" /> */}
            </form>
          </div>
        }
      </div>
    )
  }
}

Chat.propTypes = {
  room: PropTypes.object,
  socket: PropTypes.object
}