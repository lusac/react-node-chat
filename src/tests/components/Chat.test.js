/* global it describe expect beforeEach */
import React from 'react';
import { shallow, configure } from 'enzyme';
import renderer from 'react-test-renderer'
import Adapter from 'enzyme-adapter-react-16';
import Chat from '../../components/Chat'

configure({ adapter: new Adapter() })

describe('Chat component', () => {
  describe('snapshot', () => {
    it('room not selected', () => {
      const room = {}
      const renderedValue =  renderer.create(<Chat room={room} />).toJSON()
      expect(renderedValue).toMatchSnapshot();
    })

    it('room selected', () => {
      const room = { id: 1, msgs: [] }
      const renderedValue =  renderer.create(<Chat room={room} />).toJSON()
      expect(renderedValue).toMatchSnapshot()
    })
  })

  describe('rendered', () => {
    describe('room not selected', () => {
      let wrapper, room = {}

      beforeEach(() => {
        wrapper = shallow(<Chat room={room}/>)
      })

      it('correct label', () => {
        expect(wrapper.find('span').get(0).props.children).toBe('Nenhum chat selecionado')
      });
    })

    describe('room selected', () => {
      let wrapper, room = { id: 1, msgs: ['mensagens chat 1'] }

      beforeEach(() => {
        wrapper = shallow(<Chat room={room}/>)
      })

      it('correct label', () => {
        expect(wrapper.find('.chat-room__messages div').get(0).props.children).toBe('mensagens chat 1')
      });
    })
  })
})