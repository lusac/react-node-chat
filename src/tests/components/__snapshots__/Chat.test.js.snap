// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat component snapshot room not selected 1`] = `
<div
  className="chat-room"
>
  <div
    className="chat-room__messages"
  >
    <span>
      Nenhum chat selecionado
    </span>
  </div>
</div>
`;

exports[`Chat component snapshot room selected 1`] = `
<div
  className="chat-room"
>
  <div
    className="chat-room__messages"
  />
  <div
    className="chat-room__input__wrapper"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        className="chat-room__input"
        onChange={[Function]}
        placeholder="escreva uma mensagem"
        value=""
      />
    </form>
  </div>
</div>
`;
